sdk/search/Azure.Search/src/Serialization/JsonExtensions.cs

```diff
@@ -0,0 +1,473 @@
+                float x when float.IsNaN(x) => Constants.NanValue,
+                float x => x.ToString(formatProvider).ToLowerInvariant()
+            };
+
+        /// <summary>
+        /// Format floating point values.
+        /// </summary>
+        /// <param name="value">Double.</param>
+        /// <param name="formatProvider">Format Provider.</param>
+        /// <returns>OData string.</returns>
+        public static string Double(double value, IFormatProvider formatProvider) =>
+            value switch
+            {
+                double.NegativeInfinity => Constants.NegativeInfValue,
+                double.PositiveInfinity => Constants.InfValue,
+                double x when double.IsNaN(x) => Constants.NanValue,
+                double x => x.ToString(formatProvider).ToLowerInvariant()
+            };
+
+        /// <summary>
+        /// Format dates.
+        /// </summary>
+        /// <param name="value">Date</param>
+        /// <param name="formatProvider">Format Provider.</param>
+        /// <returns>OData string.</returns>
+        public static string Date(DateTime value, IFormatProvider formatProvider) =>
+            Date(
+                value.Kind == DateTimeKind.Unspecified ?
+                    new DateTimeOffset(value, TimeSpan.Zero) :
+                    new DateTimeOffset(value),
+                formatProvider);
+
+        /// <summary>
+        /// Format dates.
+        /// </summary>
+        /// <param name="value">Date</param>
+        /// <param name="formatProvider">Format Provider.</param>
+        /// <returns>OData string.</returns>
+        public static string Date(DateTimeOffset value, IFormatProvider formatProvider) =>
+            value.ToString("o", formatProvider);
+
+        /// <summary>
+        /// Parse JSON into a SearchDocument.
+        /// </summary>
+        /// <param name="reader">The JSON reader.</param>
+        /// <param name="typeToConvert">The type to convert to.</param>
+        /// <param name="options">Serialization options.</param>
+        /// <returns>A deserialized SearchDocument.</returns>
+        public static SearchDocument ReadSearchDocument(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options)
+        {
+            Debug.Assert(typeToConvert != null);
+            Debug.Assert(typeToConvert.IsAssignableFrom(typeof(SearchDocument)));
+
+            SearchDocument doc = new SearchDocument();
+            reader.Expects(JsonTokenType.StartObject);
+            while (reader.Read())
+            {
+                if (reader.TokenType == JsonTokenType.EndObject) { break; }
+
+                string propertyName = reader.ExpectsPropertyName();
+
+                // Ignore OData properties - we don't expose those on custom
+                // user schemas
+                if (!propertyName.StartsWith(Constants.ODataKeyPrefix, StringComparison.OrdinalIgnoreCase))
```

Comment:
The OData control properties shouldn't get customized by the JsonSerializerOptions.  That said, this is just the string `"@"` so Ordinal would also be fine.

Reply:
OK. Either way we choose, if the user passes in different options, it might conflict with user expectations. I am not well-versed in the scenario to know if this is OK or not, so just wanted to bring it up. Looks like it has been thought through :)

Reply:
Thanks for the heads up - that'll be interesting.